File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/InviteVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * The shared invite table (spec/invite-vectors.json), run vector by vector so a rule that drifts on
 * one platform reddens here and nowhere else. Every vector id is asserted, so a table the test stops
 * reading is a failure, not silence.
 */
class InviteVectorsTest {
    private fun vectorsFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "spec/invite-vectors.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        error("spec/invite-vectors.json not found above ${File(".").absolutePath}")
    }

    @Suppress("UNCHECKED_CAST")
    private fun table(name: String): List<Map<String, Any?>> {
        val root = FlareJson.parse(vectorsFile().readText()) as Map<String, Any?>
        return root.getValue(name) as List<Map<String, Any?>>
    }

    private fun mode(name: Any?): FlareInviteCodeMode = when (name) {
        "off" -> FlareInviteCodeMode.Off
        "required" -> FlareInviteCodeMode.Required
        else -> FlareInviteCodeMode.Optional
    }

    @Suppress("UNCHECKED_CAST")
    private fun result(raw: Any?): FlareInviteCodeCheckResult? = (raw as? Map<String, Any?>)?.let {
        FlareInviteCodeCheckResult(it["valid"] as Boolean, it["inviterDisplayName"] as? String)
    }

    @Suppress("UNCHECKED_CAST")
    private fun stats(raw: Any?): FlareReferralStats? = (raw as? Map<String, Any?>)?.let {
        FlareReferralStats((it["direct"] as Number).toInt(), (it["l2"] as Number).toInt(), (it["l3"] as Number).toInt(), (it["total"] as Number).toInt())
    }

    private fun length(v: Map<String, Any?>) = (v["length"] as? Number)?.toInt() ?: FLARE_INVITE_CODE_DEFAULT_LENGTH

    @Test fun normalizesLikeTheTableAndIdempotently() {
        val rows = table("normalize")
        assertEquals(10, rows.size)
        for (v in rows) {
            val once = normalizeInviteCode(v["raw"] as String, length(v))
            assertEquals(v["expected"], once, v["id"].toString())
            assertEquals(once, normalizeInviteCode(once, length(v)), "${v["id"]} idempotent")
        }
    }

    @Test fun resolvesTheFieldState() {
        val rows = table("fieldState")
        assertEquals(13, rows.size)
        for (v in rows) {
            val state = inviteCodeFieldState(
                mode = mode(v["mode"]), value = v["value"] as String,
                checking = v["checking"] as? Boolean ?: false, checkResult = result(v["checkResult"]),
                error = v["error"] as? String, disabled = v["disabled"] as? Boolean ?: false,
            )
            assertEquals(v["expected"], state.name.lowercase(), v["id"].toString())
        }
    }

    @Test fun asksForACheckOnlyForACompleteEnabledCode() {
        val rows = table("checkRequest")
        assertEquals(6, rows.size)
        for (v in rows) {
            val code = inviteCodeToCheck(v["value"] as String, length(v), mode(v["mode"]), v["disabled"] as? Boolean ?: false)
            assertEquals(v["expected"], code, v["id"].toString())
        }
    }

    @Suppress("UNCHECKED_CAST")
    @Test fun listsDepthRowsPerVisibility() {
        val rows = table("depthRows")
        assertEquals(6, rows.size)
        for (v in rows) {
            val actual = referralDepthRows(stats(v["stats"]), (v["maxDepthShown"] as Number).toInt()).map { it.name.lowercase() }
            assertEquals(v["expected"] as List<String>, actual, v["id"].toString())
        }
    }

    @Suppress("UNCHECKED_CAST")
    @Test fun resolvesRegenerateCooldownRoundingUp() {
        val rows = table("regenerate")
        assertEquals(10, rows.size)
        for (v in rows) {
            val a = regenerateAvailability(v["canRegenerate"] as Boolean, (v["availableAt"] as? Number)?.toLong(), (v["now"] as Number).toLong())
            val expected = v["expected"] as Map<String, Any?>
            assertEquals(expected["shown"], a.shown, "${v["id"]} shown")
            assertEquals(expected["enabled"], a.enabled, "${v["id"]} enabled")
            val remaining = expected["remaining"] as? Map<String, Any?>
            if (remaining == null) assertNull(a.remaining, v["id"].toString())
            else {
                assertEquals(remaining["unit"], a.remaining!!.unit.name.lowercase(), "${v["id"]} unit")
                assertEquals((remaining["count"] as Number).toInt(), a.remaining!!.count, "${v["id"]} count")
            }
        }
    }

    @Test fun formatsJoinedDates() {
        val rows = table("joinedDate")
        assertEquals(2, rows.size)
        for (v in rows) {
            assertEquals(v["expected"], formatInviteJoinedDate((v["year"] as Number).toInt(), (v["month"] as Number).toInt(), (v["day"] as Number).toInt()), v["id"].toString())
        }
    }

    @Test fun cooldownAndDepthCopyComeFromTheStringsTable() {
        val s = FlareStrings()
        assertEquals("15 分钟 后可重新生成", inviteCooldownText(s, FlareInviteCooldown(FlareInviteCooldownUnit.Minute, 15)))
        assertEquals("2 小时 后可重新生成", inviteCooldownText(s, FlareInviteCooldown(FlareInviteCooldownUnit.Hour, 2)))
        assertEquals("团队总数", inviteDepthLabel(s, FlareReferralDepth.Total))
        val en = FlareStrings { myInviteCooldown = "You can regenerate in {time}"; myInviteUnitDays = "{n} d" }
        assertEquals("You can regenerate in 3 d", inviteCooldownText(en, FlareInviteCooldown(FlareInviteCooldownUnit.Day, 3)))
    }
}

File: packages/flutter-im-ui/test/invite_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/src/models/invite_data.dart';
import 'package:flutter_test/flutter_test.dart';

// The shared invite table (spec/invite-vectors.json), run vector by vector so a
// rule that drifts on one platform reddens here and nowhere else.
void main() {
  final file = File('${Directory.current.path}/../../spec/invite-vectors.json');
  final data = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  List<Map<String, dynamic>> table(String key) =>
      (data[key] as List).cast<Map<String, dynamic>>();

  FlareInviteCodeMode mode(String? name) => switch (name) {
    'off' => FlareInviteCodeMode.off,
    'required' => FlareInviteCodeMode.required,
    _ => FlareInviteCodeMode.optional,
  };
  FlareInviteCodeCheckResult? result(Map<String, dynamic>? raw) => raw == null
      ? null
      : FlareInviteCodeCheckResult(
          valid: raw['valid'] as bool,
          inviterDisplayName: raw['inviterDisplayName'] as String?,
        );
  FlareReferralStats? stats(Map<String, dynamic>? raw) => raw == null
      ? null
      : FlareReferralStats(
          direct: raw['direct'] as int,
          l2: raw['l2'] as int,
          l3: raw['l3'] as int,
          total: raw['total'] as int,
        );

  group('normalize', () {
    for (final v in table('normalize')) {
      test(v['id'] as String, () {
        final length = (v['length'] as int?) ?? flareInviteCodeDefaultLength;
        final once = normalizeInviteCode(v['raw'] as String, length);
        expect(once, v['expected']);
        expect(normalizeInviteCode(once, length), once, reason: 'idempotent');
      });
    }
  });

  group('fieldState', () {
    for (final v in table('fieldState')) {
      test(v['id'] as String, () {
        final state = inviteCodeFieldState(
          mode: mode(v['mode'] as String?),
          value: v['value'] as String,
          checking: (v['checking'] as bool?) ?? false,
          checkResult: result(v['checkResult'] as Map<String, dynamic>?),
          error: v['error'] as String?,
          disabled: (v['disabled'] as bool?) ?? false,
        );
        expect(state.name, v['expected']);
      });
    }
  });

  group('checkRequest', () {
    for (final v in table('checkRequest')) {
      test(v['id'] as String, () {
        expect(
          inviteCodeToCheck(
            value: v['value'] as String,
            length: (v['length'] as int?) ?? flareInviteCodeDefaultLength,
            mode: mode(v['mode'] as String?),
            disabled: (v['disabled'] as bool?) ?? false,
          ),
          v['expected'],
        );
      });
    }
  });

  group('depthRows', () {
    for (final v in table('depthRows')) {
      test(v['id'] as String, () {
        final rows = referralDepthRows(
          stats(v['stats'] as Map<String, dynamic>?),
          v['maxDepthShown'] as int,
        );
        expect(rows.map((r) => r.name).toList(), v['expected']);
      });
    }
  });

  group('regenerate', () {
    for (final v in table('regenerate')) {
      test(v['id'] as String, () {
        final a = regenerateAvailability(
          v['canRegenerate'] as bool,
          v['availableAt'] as int?,
          v['now'] as int,
        );
        final expected = v['expected'] as Map<String, dynamic>;
        expect(a.shown, expected['shown']);
        expect(a.enabled, expected['enabled']);
        final remaining = expected['remaining'] as Map<String, dynamic>?;
        if (remaining == null) {
          expect(a.remaining, isNull);
        } else {
          expect(a.remaining!.unit.name, remaining['unit']);
          expect(a.remaining!.count, remaining['count']);
        }
      });
    }
  });

  group('joinedDate', () {
    for (final v in table('joinedDate')) {
      test(v['id'] as String, () {
        expect(
          formatInviteJoinedDate(v['year'] as int, v['month'] as int, v['day'] as int),
          v['expected'],
        );
      });
    }
  });
}
